spring-integration-zmq/src/main/java/org/springframework/integration/zmq/inbound/ZmqMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,326 @@
+						this.consumerStopAction.equals(ConsumerStopAction.UNSUBSCRIBE_CLEAN) &&
+						this.cleanSession) {
+					this.client.unsubscribe(getTopic().getBytes(Charset.defaultCharset()));
+				}
+			}
+			catch (ZMQException e) {
+				logger.error("Exception while unsubscribing", e);
+			}
+			try {
+				if (getUrl() != null) {
+					this.client.disconnect(getUrl());
+				}
+				else {
+					this.client.disconnect(this.clientFactory.getServerURI());
+				}
+			}
+			catch (ZMQException e) {
+				logger.error("Exception while disconnecting", e);
+			}
+			try {
+				this.poller.close();
+				this.client.close();
+				this.context.term();
+			}
+			catch (ZMQException e) {
+				logger.error("Exception while closing", e);
+			}
+			this.connected = false;
+			this.client = null;
+		}
+	}
+
+	@Override
+	public void setTopic(String topic) {
+		this.topicLock.lock();
+		try {
+			super.setTopic(topic);
+			if (this.client != null) {
+				this.client.subscribe(topic.getBytes(Charset.defaultCharset()));
+			}
+		}
+		catch (ZMQException e) {
+			super.removeTopic(topic);
+			throw new MessagingException("Failed to subscribe to topic " + topic, e);
+		}
+		finally {
+			this.topicLock.unlock();
+		}
+	}
+
+	@Override
+	public void removeTopic(String topic) {
+		this.topicLock.lock();
+		try {
+			if (this.client != null) {
+				this.client.unsubscribe(topic.getBytes(Charset.defaultCharset()));
+				if (this.clientFactory.getClientType() == ZMQ.SUB) {
+					this.client.subscribe("".getBytes(Charset.defaultCharset()));
+				}
+			}
+			super.removeTopic(topic);
+		}
+		catch (ZMQException e) {
+			throw new MessagingException("Failed to unsubscribe from topic " + topic, e);
+		}
+		finally {
+			this.topicLock.unlock();
+		}
+	}
+
+	private synchronized void connectAndSubscribe() {
+		this.cleanSession = this.clientFactory.cleanSession();
+		this.consumerStopAction = this.clientFactory.getConsumerStopAction();
+		if (this.consumerStopAction == null) {
+			this.consumerStopAction = ConsumerStopAction.UNSUBSCRIBE_CLEAN;
+		}
+		Assert.state(getUrl() != null || this.clientFactory.getServerURI() != null,
+				"If no 'url' provided, clientFactory.getServerURI() must not be null");
+		this.context = this.clientFactory.getContext();
+		this.client = this.clientFactory.getClientInstance(getClientId(), getTopic());
+
+		if (this.clientFactory.getUserName() != null && this.clientFactory.getPassword() != null) {
+			this.client.setPlainUsername(this.clientFactory.getUserName().getBytes(Charset.defaultCharset()));
+			this.client.setPlainPassword(this.clientFactory.getPassword().getBytes(Charset.defaultCharset()));
+		}
+
+		this.topicLock.lock();
+		String topic = getTopic();
+
+		try {
+			if (getUrl() != null) {
+				this.client.connect(getUrl());
+			}
+			else {
+				this.client.connect(this.clientFactory.getServerURI());
+			}
+
+			this.poller = this.clientFactory.getPollerInstance(ZMQ.Poller.POLLIN);
+		}
+		catch (ZMQException | NullPointerException e) {
```

Comment:
`NullPointerException`?
What made you to add this here?
May be need to change the logic and avoid such a catching?